docs/specs/6-auth-via-aad.md

```diff
@@ -0,0 +1,95 @@
+
+In the local auth workflow, we use a permanent `AccessKey` to sign `ClientToken`, which the client uses the `ClientToken` to connect to our service.
+
+Though the `AccessKey` can be regenerated, different service principals share the same `AccessKey` may lead to security issues. And the leak of `AccessKey` (customers have to store the key in their machine or config) is also a problem.
+
+## Supported scenarios
+
+- Allow Web PubSub service to auth API requests via Azure AD token.
+
+- Allow clients to connect to Web PubSub Service hub without using local auth methods.
+
+## Development Plan
+
+- [ ] **Phase 1** Support C# SDK
+
+- [ ] **Phase 2** Support other languages (python/java/golang, etc.)
+
+## Workflow
+
+![functions workflow](../images/auth-via-aad.png)
+
+1. Client sends a negotiate request to App to get a `ClientToken`.
+
+2. App calls [MSAL library](https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-overview), with identity info in the connection string.
+
+3. App receives the `AAD Token` from Azure authentication endpoint.
+
+    _The ADD Token contains identity info of the service principal such as objectId, tenantId, groups._
+
+4. App calls GET `ClientToken` API, align with the `AAD Token`.
+
+5. Service sends a request to Azure resource manager endpoint _(rbac)_.
+
+6. Service receive permitted actions.
+
+7. Service check if the service principal had the _clientToken/action_ permission. If so, get or generate a temporary `AccessKey`, then sign a `ClientToken` with the `AccessKey`.
+
+    _The key id of AccessKey will be associated with the objectId, which means we have the ability to know from which service principal the ClientToken has been signed._
+
+8. App receives the `ClientToken`.
+
+9. Client receives the `ClientToken`.
+
+10. Client creates WebSocket connection with the `ClientToken`.
+
+11. Service check if the `ClientToken` was signed by known `AccessKeys` and check if the service principal had the _clientConection/write_ permission.
+
+12. Client receive 101 or 401 response.
+
+## API design
+
+### GET /auth/clientToken
+
+#### URL
+
+`https://<service endpoint>/api/v1/auth/clientToken`
+
+#### Request
+
+| Name | Type | Required | Sample | Description |
+|---|---|---|---|---|
+| `hub` | string | \* | chat | The hub name |
+| `userId` | string | | foo | The userId for the client |
+| `roles` | string | | webpubsub.joinLeaveGroup,webpubsub.sendToGroup | The roles that the client will have, separated by comma. |
+| `ttl` | integer | | 5 | The time-to-live minutes for the access token. If not set, the default value is 60 minutes. |
+
+#### Request Sample
+
+```url
+GET https://<endpoint>/api/v1/auth/clientToken?hub=chat&userId=foo&roles=webpubsub.joinLeaveGroup,webpubsub.sendToGroup
+```
+
+#### Response
+
+Response will be a json object, which contains following properties.
+
+```json
+{
+    "ClientToken": "<JWT Token>, see https://jwt.io for detailed information.",

```

Comment:
keep the format consistent with the sdk to use `Token` as the property name?

Reply:
We also need to provide the subpath, e.g. `/client/hub/chat`..
Shall we directly provide the full URL, containing host name, subpath, and query string?